pymc_marketing/prior.py

```diff
@@ -0,0 +1,893 @@
+
+        """
+
+        if self.transform:
+            raise ValueError("Can't constrain a transformed variable")
+
+        new_parameters = pm.find_constrained_prior(
+            self.pymc_distribution,
+            lower=lower,
+            upper=upper,
+            init_guess=self.parameters,
+            **kwargs,
+        )
+
+        return Prior(
+            self.distribution,
+            dims=self.dims,
+            transform=self.transform,
+            centered=self.centered,
+            **new_parameters,
+        )
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, Prior):
+            return False
+
+        try:
+            np.testing.assert_equal(self.parameters, other.parameters)
+        except AssertionError:
+            return False
+
+        return (
+            self.distribution == other.distribution
+            and self.dims == other.dims
+            and self.centered == other.centered
+            and self.transform == other.transform
+        )
+
+    def sample_prior(
+        self, coords=None, name: str = "var", **sample_prior_predictive_kwargs
+    ) -> xr.Dataset:
+        """Sample the prior distribution for the variable.
+
+        Parameters
+        ----------
+        coords : dict[str, list[str]], optional
+            The coordinates for the variable, by default None.
+            Only required if the dims are specified.
+        name : str, optional
+            The name of the variable, by default "var".
+        sample_prior_predictive_kwargs : dict
+            Additional arguments to pass to `pm.sample_prior_predictive`.
+
+        Returns
+        -------
+        xr.Dataset
+            The dataset of the prior samples.
+
+        Example
+        -------
+        Sample from a hierarchical normal distribution.
+
+        .. code-block:: python
+
+            dist = Prior(
+                "Normal",
+                mu=Prior("Normal"),
+                sigma=Prior("HalfNormal"),
+                dims="channel",
+            )
+
+            coords = {"channel": ["C1", "C2", "C3"]}
+            prior = dist.sample_prior(coords=coords)
+
+        """
+        coords = coords or {}
+
+        if missing_keys := set(self.dims) - set(coords.keys()):
+            raise KeyError(f"Coords are missing the following dims: {missing_keys}")
+
+        with pm.Model(coords=coords):
+            self.create_variable(name)
+
+            return pm.sample_prior_predictive(**sample_prior_predictive_kwargs).prior
+
+    def __deepcopy__(self, memo) -> Prior:
+        if id(self) in memo:
+            return memo[id(self)]
+
+        copy_obj = Prior(
+            self.distribution,
+            dims=copy.copy(self.dims),
+            centered=self.centered,
+            transform=self.transform,
+            **copy.deepcopy(self.parameters),
+        )
+        memo[id(self)] = copy_obj
+        return copy_obj
+
+    def copy(self) -> Prior:
```

Comment:
shall we simply call this `decopy`?

Reply:
How about `deepcopy` specifically then? 
I read that as de-copy